File: Python/LeetCode/44.Wildcard_matching.py
# Wildcard matching
class Solution:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        if not p:
            return not s
        if len(p) == 1:
            return s == p or p == '?'
        if p[1] != '*':
            return (s and (s[0] == p[0] or p[0] == '?') and self.isMatch(s[1:], p[1:]))
        else:
            return self.isMatch(s, p[2:]) or (s and (s[0] == p[0] or p[0] == '?') and self.isMatch(s[1:], p))   

print(Solution().isMatch("aa", "a"))

File: Javascript/LeetCode/44.Wildcard_Matching.js
// Wildcard Matching
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    let dp = new Array(s.length + 1);
    for (let i = 0; i <= s.length; i++) {
        dp[i] = new Array(p.length + 1);
        for (let j = 0; j <= p.length; j++) {
            if (i === 0 && j === 0) {
                dp[i][j] = true;
            } else if (i === 0) {
                dp[i][j] = p[j - 1] === '*' ? dp[i][j - 1] : false;
            } else if (j === 0) {
                dp[i][j] = false;
            } else {
                if (p[j - 1] === '*') {
                    dp[i][j] = dp[i][j - 1] || dp[i - 1][j];
                } else if (p[j - 1] === '?') {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = s[i - 1] === p[j - 1] && dp[i - 1][j - 1];
                }
            }
        }
    }
    return dp[s.length][p.length];  
};

console.log(isMatch('aa', 'a'));